Rework library scan / spotify webapi integration

chme opened the pull request:
Hi @ejurgensen 

as shortly discussed, this pr refactors the scan of items into the db. It introduces a "library source" (there is probably a better name for it ...) abstraction. This should (hopefully) make it easier to add new sources in the future. For now the two existing sources "filescanner" and "spotify" are implementing this abstraction. The main changes are done in the commits up to 87e1bf9.

Starting with commit 5d80a6e i worked on the spotify integration. The main changes are:
* forked-daapd now stores the refresh-token for the webapi access in the admin table. The refresh-token allows forked-daapd to access the webapi after restarts without additional user interaction (the oauth process must only be done once)
* Metadata is read from the webapi
* Instead of reading the saved tracks, forked-daapd now reads the saved albums. This reduces the number of requests made
* The saved playlists are now read from the webapi

If forked-daapd does not have a valid acces-token for the webapi, scan is done with libspotify (hopefully without regressions). If forked-daapd has a valid access-token, libspotify is only used for playback, artwork and playlist update trigger.

There are currently some drawbacks using the webapi:
* the webapi only returns genre for (full) artist objects
* the webapi only returns a simplified album object for tracks (either through the saved tracks endpoint or the playlist tracks endpoint. This album object is missing the "release date".

I still need to run some additional tests and there are probably some things i missed. But I would appreciate if you can give this a review.

ejurgensen commented:
Sounds really great, I look forward to trying it. I noticed recently that libspotify no longer gives me the "Discover weekly" playlist, so I'm hoping this change brings it back.

I notice there are a great deal of commits. Do you think it would be possible to squash them a bit?

BTW did you ever find out why mdns forked-daapd.local wasn't working?

chme commented:
Squashed the commits a bit. Also fixed some bugs.

> I noticed recently that libspotify no longer gives me the "Discover weekly" playlist, so I'm hoping this change brings it back.

I noticed this too and can confirm that the "Discover weekly" playlist reappears with these changes.

> BTW did you ever find out why mdns forked-daapd.local wasn't working?

No, i could not resolve this (but i also did not try that hard). It might be that it is an issue with the linux distro i am using (solus).

ejurgensen commented on src/library.c:
I notice you started using bool a few places in this pr. I'm all for that, but I'm wondering how standard bool is these days? Its been years since I last attempted bool, so not sure about the state of it.

ejurgensen commented on src/filescanner.c:
```diff
@@ -62,6 +62,7 @@
 #include "cache.h"
```
I suggest placing library modules in a subdir. Like outputs (and the future inputs).

ejurgensen commented on src/logger.c:
```diff
@@ -43,7 +43,7 @@ static int threshold;
 static char *logfilename;
 static FILE *logfile;
-static char *labels[] = { "config", "daap", "db", "httpd", "http", "main", "mdns", "misc", "rsp", "scan", "xcode", "event", "remote", "dacp", "ffmpeg", "artwork", "player", "raop", "laudio", "dmap", "dbperf", "spotify", "lastfm", "cache", "mpd", "stream", "cast", "fifo" };
+static char *labels[] = { "config", "daap", "db", "httpd", "http", "main", "mdns", "misc", "rsp", "scan", "xcode", "event", "remote", "dacp", "ffmpeg", "artwork", "player", "raop", "laudio", "dmap", "dbperf", "spotify", "lastfm", "cache", "mpd", "stream", "cast", "fifo", "lib" };
```
Worth having an entire log domain for this?

chme replied:
I guess, i can also reuse the "scan" logdomain. But in a future pr (adding non library items and the spotify search) it will not be limited to the scanning of items. I am fine either way and will change the logdomain if you want to.

ejurgensen replied:
Ok let's keep it. However, you should also update forked-daapd.8 because the domains are also listed there.

ejurgensen commented:
Went through the pr and it looks very good to me. Only have a few minor comments, see above. Haven't had time for much testing yet.

ejurgensen commented on src/spotify_webapi.h:
```diff
@@ -0,0 +1,121 @@
+
+
+#include <event2/event.h>
+#ifdef HAVE_JSON_C_OLD
```
Note that when the other pr gets merged this won't be required any more

chme commented on src/library.c:
I have to say my knowledge here is limited. It is part of C99 and i assumed this would not cause issues in 2017.

chme commented:
I moved the filescanner* files to the "library" folder. And found a memory leak (the artists string was never freed). Only the first artist of a track/album is now used. I think this is what libspotify does, as there is only one artist per track/album. I will do some more tests over the coming weekend.

ejurgensen commented on src/misc.c:
```diff
@@ -261,6 +261,15 @@ safe_hextou64(const char *str, uint64_t *val)
 }
 
+char *
+safe_strdup(const char *str)
```
Hehe, the need for this is C at it's worst :-)

ejurgensen commented on src/spotify.c:
```diff
@@ -1344,15 +1002,30 @@ static void playlist_update_in_progress(sp_playlist *pl, bool done, void *userda
 
-      spotify_playlist_save(pl);
+      if (spotify_access_token_valid)
+        {
```
Is the indentation ok here? Here on github the brackets don't seem equally aligned.

ejurgensen commented:
Added a few more comments, and also got around to testing it, and so far it is working well. Very glad to have the "Discover weekly" list again. I little thing I noticed during startup was that the scan ends with this:

[2017-01-17 21:46:35] [  LOG]  spotify: Scanned 35 of 37 saved playlists (590 tracks)
[2017-01-17 21:46:35] [  LOG]  spotify: Scanned 36 of 37 saved playlists (608 tracks)
(apologies for not providing you with the debug log, but can't get that right now)

So one playlist doesn't get scanned. Could a reason be that one of my playlists is actually empty?

ejurgensen commented:
BTW, when you have looked at the comments above, let me know if you think it is ready to merge.

chme commented on src/library.c:
```diff
@@ -0,0 +1,851 @@
+#include "commands.h"
+#include "conffile.h"
+#include "db.h"
+#include "library/filescanner.h"
```
The source abstraction is not complete. Making it entirely generic requires a lot more work. Maybe some kind of metadata-scanner interface that maps a data kind to a scan function can solve this. But i am not sure who should own the scan function (would it fit into the inputs changes you are working on?). 

In my opinion, looking into this in depth should happen outside of this pr.

ejurgensen replied:
Ok, sounds sensible. The inputs I'm working on is just focused on getting audio and metadata from the stream to the player, so right now it is not a suitable owner of the scan. Anyway, I agree, let's start by getting this up and running, then we can consider later how to make it more generic.

chme commented on src/Makefile.am:
```diff
@@ -2,11 +2,11 @@
 
 if COND_SPOTIFY
-SPOTIFY_SRC=spotify.c spotify.h
+SPOTIFY_SRC=spotify.c spotify.h spotify_webapi.c spotify_webapi.h
```
The webapi-files only help access the spotify webapi, the real library source is in spotify.c. I did not move spotify.c (+ *_webapi) because they do more than only filling the library.

chme commented:
> So one playlist doesn't get scanned. Could a reason be that one of my playlists is actually empty?

That can be the reason, other reasons to skip a playlist are missing uri or name. I increased the log level for the skip message to give the user a hint which playlist was skipped.

I added a few commits to address your comments. From my side this is ready to be merged.

ejurgensen commented:
Merged, and once again thanks for a great contribution

ejurgensen commented on src/spotify_webapi.c:
```diff
@@ -0,0 +1,754 @@
+  else
+    {
+      // Next paging request
+      next_uri = strdup(request->next_uri);
```
Sorry, a few extra remarks: Why strdup'ing here and then freeing again right below? Why not just set the pointer?

chme replied:
request->next_uri is supposed to hold the string from the json response and that will be freed upon calling spotifywebapi_request_end(...). (Its not entirly true, see the other comment)

The uri needs to survive the request_end call, that is why i use strdup. It would also be possible to allocate a string big enough on the stack and copy the next_uri into it. What do you prefer?

ejurgensen replied:
Ok, I get it know. I think in this case I am indifferent to stack/heap, so that is up to you. A minor other I noticed is that there is function in spotifywebapi called "http_client_ctx_free". The prefix of that seems to indicate it is belongs to http.h, so either call it something else, or move the function to http.h/.c if it is generic in nature.

ejurgensen commented on src/spotify_webapi.c:
```diff
@@ -0,0 +1,754 @@
+  request->total = jparse_int_from_obj(request->haystack, "total");
+  tmp = jparse_str_from_obj(request->haystack, "next");
+  if (tmp)
+    request->next_uri = strdup(tmp);
```
Does this mean that if there is no tmp (next is missing), then ->next_uri will be unchanged and it will continue requesting the same page forever? Please safeguard against "ddos'ing" Spotify :-)

ejurgensen replied:
Also: Isn't request->next_uri leaking? Can't see how it is being freed.

chme replied:
The call to "request_uri(...)" sets next_uri to NULL (call to memset in line 359), if the response does not contain the element "next", spotifywebapi_request_next(...) returns early with -1.

But you are right about the leak, there is one strdup to many.

ejurgensen commented on src/spotify_webapi.c:
```diff
@@ -0,0 +1,754 @@
+
+  tmp = jparse_str_from_obj(haystack, "access_token");
+  if (tmp)
+    spotify_access_token  = strdup(tmp);
```
Possible leak if called multiple times? (yes, this one is my own mistake!)

ejurgensen commented on src/spotify_webapi.c:
```diff
@@ -0,0 +1,754 @@
+
+  tmp = jparse_str_from_obj(haystack, "refresh_token");
+  if (tmp)
+    spotify_refresh_token = strdup(tmp);
```
Same
